gapis/api/vulkan/frame_loop.go

```diff
@@ -428,18 +500,108 @@ func (f *frameLoop) resetImages(ctx context.Context, sb *stateBuilder) error {
 	for dst, src := range f.imageToBackup {
 		dstObj := s.Images().Get(dst)
 
-		primeable, err := imgPrimer.newPrimeableImageDataFromDevice(src, dst)
-		if err != nil {
-			return log.Errf(ctx, err, "Create primeable image data for image %v", dst)
+		prime := func() error {
+			primeable, err := imgPrimer.newPrimeableImageDataFromDevice(src, dst)
+			if err != nil {
+				return log.Errf(ctx, err, "Create primeable image data for image %v", dst)
+			}
+			defer primeable.free(sb)
+			err = primeable.prime(sb, useSpecifiedLayout(dstObj.Info().InitialLayout()), sameLayoutsOfImage(dstObj))
+			if err != nil {
+				return log.Errf(ctx, err, "Priming image %v with data", dst)
+			}
+			log.D(ctx, "Prime image from [%v] to [%v] succeed", src, dst)
+			return nil
 		}
-		defer primeable.free(sb)
-		err = primeable.prime(sb, useSpecifiedLayout(dstObj.Info().InitialLayout()), sameLayoutsOfImage(dstObj))
-		if err != nil {
-			return log.Errf(ctx, err, "Priming image %v with data", dst)
+
+		if err := prime(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *frameLoop) resetFences(ctx context.Context, sb *stateBuilder) error {
+	s := GetState(sb.newState)
+
+	for k := range f.fenceChanged {
```

Comment:
So this is not correct.
If the fence has been signaled (from the APIs perspective) we cannot simply reset it. We have to wait until the work that signaled it is actually done.

So we should vkWaitForFences() to wait until the fences have been actually signaled by the GPU. Remember the state mutation does not happen at the same time as the GPU, it only records the earliest it COULD be done.